### ISSUE TITLE
Crash In Leanplum SDK on start

### ISSUE THREAD

**Issue by justin** [Leanplum/Leanplum-iOS-SDK#107]
Hey,

We are seeing this crash in our app related to the Leanplum SDK that is happening when calling start.  We can't symbolicate your framework since we don't have the dSYM available, so I'm somewhat at a loss as to what to do next to give you any useful info on this. 

Crash log is at: https://gist.github.com/justin/4f31cfc7279212c2833991cc303a914a

Thanks.

**Comment by alexisoyama** [comment]
Are you using 2.0.4? If so the dSYM is at  https://storage.cloud.google.com/leanplum-sdk-release/Leanplum-2.0.4-SDK-mapping.zip

I tried to symbolicate the crash log but had no luck. Can you give it a try?

Also are you reproducing this crash on development? (not in crashlytics)

**Comment by justin** [comment]
You can reproduce it line-by-line.

`atos -arch arm64 -o ~/Development/TEDapp/Carthage/Build/iOS/Leanplum.framework/Leanplum -l 0x105820000 0xMEMORYADDRESS`

I updated my Gist above with the symbolicated version.

**Comment by alexisoyama** [comment]
Thanks! To clarify, you didn't use the dSYM to symbolicate it right? I'm in a middle of symbolicating other crash reports and this info will be helpful. 

It seems like you are getting this error in development (not in crashlytics). I need some info to reproduce this issue.
- Are you using Swift? If so, what version?
- Is this crash happening every time? or only on some occasion?
- Any code snippet that involves Leanplum will be helpful.

**Comment by alexisoyama** [comment]
Can you share the log? I believe you will get some sort of error message, like "SQLite fail to open"

**Comment by justin** [comment]
Can you clarify which log you'd like besides the [crash log](https://gist.github.com/justin/4f31cfc7279212c2833991cc303a914a)? This is coming in through Xcode's crash reporter. These are production builds that we are shipping through TestFlight.

The crash is Non-UI so it's happening in the background (see `Role:` in the crash log)

**Comment by alexisoyama** [comment]
That makes sense. In that case this is good enough. Will patch it. Thanks for the report.

**Comment by justin** [comment]
Just as an update, I checked our stats this morning after the weekend and this is our biggest crash according to the Xcode Crash Organizer by 230x compared to the second biggest.

**Comment by alexisoyama** [comment]
Jeffery is aware of the issue. Let's move the conversation there and we can work something out.
Meanwhile I'll be working on the fix.

**Comment by alexisoyama** [comment]
Fixed this in: https://github.com/Leanplum/Leanplum-iOS-SDK/pull/108

Will close this when the next SDK is released.

**Comment by justin** [comment]
Thanks for looking into this. Looking at that PR, I am not 100% sure I understand the fix. From my understanding it's happening specifically when the device is locked because our app uses the data protection APIs, so when the device is locked you can't write to the `__leanplum.sqlite` file.

Also, won't deleting the file every time the device is locked make us lose a decent chunk of analytics events?

**Comment by alexisoyama** [comment]
Good to know about the data protection APIs. What is the duration of the lock? Can you tell me more about this?

In the past when we try to init/open the database, we log this error and tries to send it to the server.This logic also uses the database and it tries to open it again, hence creating the init/open loop that caused your crash. We removed this loop.

We will delete the DB file when it is corrupted or when the DB queries fails 25 times in a row, so it will not delete it every time when the device is locked.

**Comment by justin** [comment]
TLDR: if an app has Data protection enabled as part of its provisioning profile, any file you create is going to assume that protection level by default. Apple's recommended default level is Complete, which basically locks down data if the device is locked.

I believe if you set `SQLITE_OPEN_FILEPROTECTION_COMPLETEUNTILFIRSTUSERAUTHENTICATION`  as part of calling `sqlite_open` that will allow the file to be opened and used at any time after the user first authenticates their device post-restart.

Look in `sqlite3.h` for the `SQLITE_OPEN_FILEPROTECTION_MASK` options. The [iOS Security Guide](https://www.apple.com/business/docs/iOS_Security_Guide.pdf) can explain these options more. Go to the "Encryption and Data Protection" section.

**Comment by alexisoyama** [comment]
This information is very helpful. Will make a test app and investigate the security options in depth.

**Comment by justin** [comment]
This crash got me to go a bit down the rabbit hole with our app and file protection on the files we own specifically, and I've uncovered a bit more info which makes stuff like this a bit more hairy.

Even if you set the Leanplum database to be `SQLITE_OPEN_FILEPROTECTION_COMPLETEUNTILFIRSTUSERAUTHENTICATION` or some other setting, that can be overridden by an MDM policy (think corporate provisioned devices) that enforces `SQLITE_OPEN_FILEPROTECTION_COMPLETE` or the equivalent of all files on the device. In that scenario, you'd be back to crashing, unless you are checking up front that you can access / read that file.

I honestly have no idea how to solve this. :)

**Comment by alexisoyama** [comment]
Thanks for the update! Will keep that in mind when trying to reproduce/fix it.

**Comment by alexisoyama** [comment]
The crash is fixed by https://github.com/Leanplum/Leanplum-iOS-SDK/pull/113 and https://github.com/Leanplum/Leanplum-iOS-SDK/pull/108. Tested with Data Protection. 
Fix is released in 2.0.5
